lib/Driver/Driver.cpp

```diff
@@ -884,20 +891,57 @@ Driver::parseArgStrings(ArrayRef<const char *> Args) {
   return ArgList;
 }
 
-DerivedArgList *Driver::translateInputArgs(const InputArgList &ArgList) const {
+DerivedArgList *
+Driver::translateInputAndPathArgs(const InputArgList &ArgList,
+                                  StringRef workingDirectory) const {
   DerivedArgList *DAL = new DerivedArgList(ArgList);
 
+  auto addPath = [workingDirectory, DAL](Arg *A) {
+    assert(A->getNumValues() == 1 && "multiple values not handled");
+    StringRef path = A->getValue();
+    if (workingDirectory.empty() || path == "-" ||
+        llvm::sys::path::is_absolute(path)) {
+      DAL->append(A);
+      return;
+    }
+
+    SmallString<64> fullPath{workingDirectory};
+    llvm::sys::path::append(fullPath, path);
+    unsigned index = DAL->getBaseArgs().MakeIndex(fullPath);
+    Arg *newArg = new Arg(A->getOption(), A->getSpelling(), index,
+                          DAL->getBaseArgs().getArgString(index), A);
+    DAL->AddSynthesizedArg(newArg);
+    DAL->append(newArg);
+  };
+
+  if (!workingDirectory.empty()) {
+    // Synthesize -Xcc -working-directory at the head of the argument list to
+    // ensure it is overridden by an explicit -Xcc -working-directory,
+    // although having a different working directory is probably incorrect.
+    auto xcc = Opts->getOption(options::OPT_Xcc);
+    DAL->AddSeparateArg(/*BaseArg=*/nullptr, xcc, "-working-directory");
+    DAL->AddSeparateArg(/*BaseArg=*/nullptr, xcc, workingDirectory);
```

Comment:
Please do not do this here. Put it in `addCommonFrontendOptions` instead.

Reply:
Is there somewhere I should be stashing the working directory so that it's available directly in that function, or should I get it from the argument list (and call make_absolute) again?